### PR TITLE
Fix OpcodeStack to handle propagation of taints properly in case of string concatenation in Java 11 and above

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
Instead of using StringBuffer or StringBuilder internally, Java 11 and above uses a dynamic call to makeConcatWithConstants() to append strings. Previously, `OpcodeStackDetector` did not handle the taint propagation properly in case of this dyanamic call which led to false negative such as the one described in issue [#2184](https://github.com/spotbugs/spotbugs/issues/2184). This PR fixes such issues by adding code to `OpcodeStackDetector` to handle this case as well.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Comment by baloghadamsoftware** [comment]
[No false positives](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?run=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Ttorrent%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20OpenGrok%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Bt%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20SpotBugs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Jenkins%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20Jenkins%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20SpotBugs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20OpenGrok%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20Bt%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29&newcheck=SpotBugs%20on%20Ttorrent%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2184%29) and no crashes on some open-source projects we used to test our PRs.

**Comment by baloghadamsoftware** [comment]
A side remark: I put the tests into the `java14` directory because the `java11` requires modules to set up and I failed to put module `javax.servlet.api` there. Javac claims that is is missing. I also tried to put it as a Gradle dependency but still without success.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java [hunk 1]
```diff
@@ -2800,12 +2805,74 @@ private boolean isMethodThatReturnsGivenReference(String clsName, String methodN
...
 
+        if ("makeConcatWithConstants".equals(methodName)) {
+            String[] args = new SignatureParser(signature).getArguments();
+            if (args.length == 1) {
+                Item i = getStackItem(0);
+                if (i.isServletParameterTainted()) {
+                    servletRequestParameterTainted = true;
+                }
+                Object sVal = i.getConstant();
+                if (sVal != null) {
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/OpcodeStack.java:2826]: "if not...else" is a double negative, could refactor.

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — DISMISSED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Review by KengoTODA — APPROVED** [review]
Confirmed that new test fails on the commit 8c5e1934df8610a529537c3ec3fe1ac0641ddd1f. Thanks for your contribution!

**Pull request merged**